KV i1871 - Handle timeout on remote connection

martinsumner opened the pull request:
https://github.com/basho/riak_kv/issues/1871

The `riak_kv_replrtq_snk` will now determine the work to be run after initialisation - so the initial connection time of PB clients no longer delays the startup of riak_kv.

The `riak_kv_replrtq_snk` functions called by the `riak_kv_replrtq_peer` have timeouts disabled so that a `riak_kv_replrtq_snk` process which is hanging on a connection timeout does not crash a calling `riak_kv_replrtq_peer` process.

The `riak_kv_replrrtq_peer` will now catch and ignore replies from the erlang http client.  The client catches timeouts on calls to the connection process, so that when the connection process replies it goes to the process which launched the client.

martinsumner committed:
Handle timeouts in the PB client

martinsumner committed:
Handle any delay in snk

The peer needs to call current_peers, and this may timeout if the process is held busy by a client initialisation that is timing out due to unavailability of a peer

martinsumner committed:
HTTP client may respond after timeout

The ibrowse http client  will gen_server:call on send_req with a timeout, and catch any timeout error.

This means that the client may receive an {error, reqd_timeout}, but the gen_server has not crashed and so may later send a gen_server:reply to the riak_kv_replrtq_peer.

These errors need to be picked up in handle_info to avoid noise of repeated crashes.

This is not a problem in riak_kv_replrtq_snk, as in the process we are always using the http client from within a short-lived spawned process.  Likewise in riak_kv_ttaaefs_manager the client is sued within a short-lived aae_exchange.

ThomasArts commented on src/riak_kv_replrtq_snk.erl:
```diff
@@ -236,50 +239,30 @@ set_workercount(QueueName, WorkerCount) ->
-                end,
-            Work = lists:map(MapPeerInfoFun, SnkQueuePeerInfo),
-            {ok, #state{enabled = true, work = Work}, ?INACTIVITY_TIMEOUT_MS};
+            gen_server:cast(?MODULE, initialise_work);
```
This is not the intended API for a gen_server.
I guess you want to continue with initial_work as soon as init is done.
If that is the case, rather have init return 
```
{ok, #state{), {continue, initialise_work}}
```

Then you write `handle_continue` function
```
handle_continue(initialise_work, State) ->
  %% the code of the handle cast for initialise work here
```

ThomasArts replied:
(Works in OTP 22 and above, checked that)

martinsumner replied:
Modified

ThomasArts commented on src/riak_kv_replrtq_snk.erl:
```diff
@@ -364,6 +347,32 @@ handle_call({current_peers, QueueN}, _From, State) ->
+            {SnkQueueName, Iteration, SnkW}
+        end,
+    Work = lists:map(MapPeerInfoFun, SnkQueuePeerInfo),
+    {noreply, State#state{enabled = true, work = Work}, ?INITIAL_TIMEOUT_MS};
```
You clearly don't anticipate to get prompt_work casts during initialisation. because you would drop them by assigning `work = Work` here. By using the continue strategy, you would not even run that risk to get something inbetween before being ready to handle it.

martinsumner committed:
Change to use handle_continue/2

Riak now OTP 22+ only, so ok to use feature added in OTP 21 rather than an artificial method for deferring work after init reply.

martinsumner committed:
More specific catch of HTTP client error message
